src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -181,6 +183,37 @@ public CallResult(ReadOnlyMemory<byte> output, bool? precompileSuccess, bool sho
     public interface IIsTracing { }
     public readonly struct NotTracing : IIsTracing { }
     public readonly struct IsTracing : IIsTracing { }
+
+    internal sealed class CodeLruCache
+    {
+        private const int CacheCount = 16;
+        private const int CacheMax = CacheCount - 1;
+        private readonly LruCache<ValueHash256, CodeInfo>[] _caches;
+
+        public CodeLruCache()
+        {
+            _caches = new LruCache<ValueHash256, CodeInfo>[CacheCount];
+            for (int i = 0; i < _caches.Length; i++)
+            {
+                // Cache per nibble to reduce contention as TxPool is very parallel
+                _caches[i] = new LruCache<ValueHash256, CodeInfo>(MemoryAllowance.CodeCacheSize / CacheCount, MemoryAllowance.CodeCacheSize / CacheCount, $"VM bytecodes {i}");
+            }
+        }
+
+        public CodeInfo Get(in ValueHash256 codeHash)
+        {
+            var cache = _caches[GetCacheIndex(codeHash)];
+            return cache.Get(codeHash);
+        }
+
+        public bool Set(in ValueHash256 codeHash, CodeInfo codeInfo)
+        {
+            var cache = _caches[GetCacheIndex(codeHash)];
+            return cache.Set(codeHash, codeInfo);
+        }
+
+        private static int GetCacheIndex(in ValueHash256 codeHash) => codeHash.Bytes[^1] & CacheMax;
```

Comment:
why ^1 and not 0?

Reply:
Disprotianate amount of contracts and addresses start with zeros; are gpu programs to hunt for the highest numbers of leading zeros due to calldata reductions:

https://github.com/johguse/ERADICATE2
https://github.com/1inch/profanity2

People don't search for the last nibbles so the distribution is fairly random